fedimint-server/src/multiplexed.rs

```diff
@@ -66,20 +59,99 @@ struct ModuleMultiplexerInner<MuxKey, Msg> {
+    /// Sender of receive callbacks
+    receive_callbacks_tx: Sender<Callback<MuxKey, Msg>>,
+    /// Sender of peer bans
+    peer_bans_tx: Sender<PeerId>,
 }
 
+type Callback<MuxKey, Msg> = (MuxKey, oneshot::Sender<(PeerId, Msg)>);
+
 impl<MuxKey, Msg> PeerConnectionMultiplexer<MuxKey, Msg>
 where
-    Msg: Serialize + DeserializeOwned + Unpin + Send + Debug,
-    MuxKey: Serialize + DeserializeOwned + Unpin + Send + Debug + Eq + Hash + Clone,
+    Msg: Serialize + DeserializeOwned + Unpin + Send + Debug + 'static,
+    MuxKey: Serialize + DeserializeOwned + Unpin + Send + Debug + Eq + Hash + Clone + 'static,
 {
     pub fn new(connections: PeerConnections<ModuleMultiplexed<MuxKey, Msg>>) -> Self {
+        let (send_requests_tx, send_requests_rx) = channel(1000);
+        let (receive_callbacks_tx, receive_callbacks_rx) = channel(1000);
+        let (peer_bans_tx, peer_bans_rx) = channel(1000);
+
+        tokio::spawn(Self::run(
+            connections,
+            Default::default(),
+            send_requests_rx,
+            receive_callbacks_rx,
+            peer_bans_rx,
+        ));
+
         Self {
-            inner: Arc::new(ModuleMultiplexerInner {
-                connections: Mutex::new(connections),
-                out_of_order: Default::default(),
-            }),
+            send_requests_tx,
+            receive_callbacks_tx,
+            peer_bans_tx,
+        }
+    }
+
+    async fn run(
+        mut connections: PeerConnections<ModuleMultiplexed<MuxKey, Msg>>,
+        mut out_of_order: ModuleMultiplexerOutOfOrder<MuxKey, Msg>,
+        mut send_requests_rx: Receiver<(Vec<PeerId>, MuxKey, Msg)>,
+        mut receive_callbacks_rx: Receiver<Callback<MuxKey, Msg>>,
+        mut peer_bans_rx: Receiver<PeerId>,
+    ) -> Cancellable<()> {
+        loop {
+            let mut key_inserted: Option<MuxKey> = None;
+            tokio::select! {
+                 // Send requests are forwarded to underlying connections
+                 send_request = send_requests_rx.recv() => {
+                    let (peers, key, msg) = send_request.ok_or(Cancelled)?;
+                    connections.send(&peers, ModuleMultiplexed { key, msg }).await?;
+                }
+                // Ban requests are forwarded to underlying connections
+                peer_ban = peer_bans_rx.recv() => {
+                    let peer = peer_ban.ok_or(Cancelled)?;
+                    connections.ban_peer(peer).await;
+                }
+                // Receive callbacks are added to callback queue by key
+                receive_callback = receive_callbacks_rx.recv() => {
+                    let (key, callback) = receive_callback.ok_or(Cancelled)?;
+                    out_of_order.callbacks.entry(key.clone()).or_default().push_back(callback);
+                    key_inserted = Some(key);
+                }
+                // Actual received messages are added message queue by key
+                receive = connections.receive() => {
```

Comment:
Since `connections.receive()` is blocking, this treat will do absolutely nothing until next message from the peer is received. This includes sending any outgoing messages to that peer. This can lead to subtle hangs.